odo/backends/dask_array.py

```diff
@@ -0,0 +1,74 @@
+from __future__ import absolute_import, division, print_function
+
+import numpy as np
+from datashape import discover
+from toolz import merge, accumulate
+from datashape.dispatch import dispatch
+from datashape import DataShape
+from operator import add
+import itertools
+
+from dask.array.core import rec_concatenate, Array, getem, get, names, from_array
+from dask.core import flatten
+from dask.compatibility import long
+
+from ..append import append
+from ..convert import convert
+from ..resource import resource
+
+@discover.register(Array)
+def discover_dask_array(a, **kwargs):
+    block = a._get_block(*([0] * a.ndim))
+    return DataShape(*(a.shape + (discover(block).measure,)))
+
+
+arrays = [np.ndarray]
+try:
+    import h5py
+    arrays.append(h5py.Dataset)
+
+    @dispatch(h5py.Dataset, (int, long))
+    def resize(x, size):
+        s = list(x.shape)
+        s[0] = size
+        return resize(x, tuple(s))
+
+    @dispatch(h5py.Dataset, tuple)
+    def resize(x, shape):
+        return x.resize(shape)
+except ImportError:
+    pass
+try:
+    import bcolz
+    arrays.append(bcolz.carray)
+
+    @dispatch(bcolz.carray, (int, long))
+    def resize(x, size):
+        return x.resize(size)
+except ImportError:
+    pass
+
+
+
+@convert.register(Array, tuple(arrays), cost=0.01)
+def array_to_dask(x, name=None, blockshape=None, **kwargs):
+    return from_array(x, blockshape=blockshape, name=name, **kwargs)
+
+
+@convert.register(np.ndarray, Array, cost=0.5)
+def dask_to_numpy(x, **kwargs):
+    return rec_concatenate(get(x.dask, x._keys(), **kwargs))
```

Comment:
I think that that would be a fascinating way to study the performance of the ecosystem.  There are probably a few things like that that we're well poised to do with odo/blaze that we haven't touched at all yet.

For the immediate purpose of changing the behavior of odo my intuition says that there are probably a few improvements to be made, but nothing groundbreaking.  The graph is sufficiently sparse that there are only usually a few sensible routes and relative costs between them usually differ by orders of magnitude, making the right choice fairly obvious (usually).  There have only been a handful of cases where we've been highly sensitive to algorithm cost.  It would be interesting to see how well we do with an unweighted graph; I'll bet it's not terrible.

For the moment I would just make sure that the 

    cost(np <- dask.array <- h5py.Dataset) > cost(np <- h5py.Dataset)

Probably this cost should be bumped up to 2 or something.